File: leetcode/algorithms/589_n_ary_tree_preorder_traversal/main.py
from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution:
    def preorder(self, root: "Node") -> List[int]:
        return self.getVal(root, [])

    def getVal(self, node: "Node", list: List[int]) -> List[int]:
        if node != None:
            list.append(node.val)
            for child in node.children:
                self.getVal(child, list)
        return list


    # Best Solution
    # Best Solution 1: Iterative
    def bestSolution1(self, root: "Node", output: List[int] = []) -> List[int]:
        if root is None:
            return []

        stack = [root]
        output = []

        # Till there is element in stack the loop runs.
        while stack:
            # pop the last element from the stack and store it into temp.
            temp = stack.pop()

            # append. the value of temp to output
            output.append(temp.val)

            # add the children of the temp into the stack in reverse order.
            # children of 1 = [3,2,4], if not reveresed then 4 will be popped out first from the stack.
            # if reversed then stack = [4,2,3]. Here 3 will pop out first.
            # This continues till the stack is empty.
            stack.extend(temp.children[::-1])

        # return the output
        return output

    # Best Solution 2: Recursive
    def bestSolution2(self, root: "Node") -> List[int]:
        output = []

        # perform dfs on the root and get the output stack
        self.dfs(root, output)

        # return the output of all the nodes.
        return output

    def dfs(self, root: "Node", output: List[int]):
        # If root is none return
        if root is None:
            return

        # for preorder we first add the root val
        output.append(root.val)

        # Then add all the children to the output
        for child in root.children:
            self.dfs(child, output)

File: leetcode/algorithms/589_n_ary_tree_preorder_traversal/main.js
function Node(val, children) {
  this.val = val;
  this.children = children;
}

/**
 * @param {Node|null} root
 * @return {number[]}
 */
var preorder = function (root) {
  const result = [];

  /**
   * @param {Node|null} node
   * @return {void}
   */
  function getVal(node) {
    if (!node) {
      return;
    }
    result.push(node.val);
    for (const child of node.children) {
      getVal(child);
    }
  }

  getVal(root);

  return result;
};


// Solution
// Solution 1
var solution1 = function (root, ans = []) {
  if (!root) {
    return ans;
  }

  ans.push(root.val);

  for (let child of root.children) {
    solution1(child, ans);
  }

  return ans;
};

// Solution 2
var solution2 = function (root) {
  const result = [];

  function traverse(tree) {
    if (!tree) {
      return result;
    }
    result.push(tree.val);
    for (let i = 0; i < tree.children.length; i++) {
      traverse(tree.children[i]);
    }
    return result;
  }

  return traverse(root);
};

// Solution 3
var solution3 = function (root) {
  const stack = [];
  const result = [];

  if (!root) {
    return result;
  }

  stack.push(root);

  while (stack.length) {
    const node = stack.pop();
    result.push(node.val);
    for (let i = node.children.length - 1; i >= 0; i--) {
      stack.push(node.children[i]);
    }
  }

  return result;
};
